src/hotspot/share/code/nmethod.cpp

```diff
@@ -1464,10 +1464,14 @@ void nmethod::purge(bool free_code_cache_data) {
     ec = next;
   }
 
-  Universe::heap()->unregister_nmethod(this);
+  if (unregister_nmethod) {
+    assert(!free_code_cache_data, "must not free when postponing unregistering");
+    Universe::heap()->unregister_nmethod(this);
```

Comment:
Could the impl of `unregister_nmethod` be `ShouldNotReachHere()` for GCs using bulk-removal?

Reply:
No, for the reasons above, we need to keep the functionality to unregister individual nmethods/code blobs for the reasons above.